### DISCUSSION TITLE
Multiple receivers support

### DISCUSSION THREAD

**Discussion by Neco777** [Arduino-IRremote/Arduino-IRremote#1074] [Show and tell]
Hello,

I worked on my project for arduino and I used IRremote library for receive IR signals. I found that the library did not support several IR receivers (connected to different pins) on a single Arduino.
The root cause of this (as far as I understood) is variable "volatile irparams_t irparams" in IRremote.cpp and "ISR(TIMER2_OVF_vect)" procedure. I made some minor changes to make it working for several receivers/pins. Now it seems to me working and I would like to see these changes in the master branch, so I can have all updates of the library.

I am not familiar with the procedure that I should follow to commit changes - I need to spend some time to figure out how to do this. But I am not sure if my changes will be approved by authors. My main concerns are:
- I am not a confident C++ coder. I came from C# world and my C++ might be awful for real masters.
- I used my own class CppList for improving IRremote library. It makes usage of IRremote more complicated for other people (I use CppList anyway so adding this #include does not bother me). So I need your opinion on that.

In short my approach is the following. I moved "irparams_t irparams" from global state to class IRrecv. In global state I created collection of all instances of irparams_t instead. I add new instance of irparams_t into that collection in every constructor call of IRrecv (and delete in destructor call). So every instance of IRrecv works with its own instance of irparams_t. Inside "ISR(TIMER2_OVF_vect)" I go thru collection of irparams_t and process values from each pin separately.

I can send you all code (if you give me your email address). If you find my approach is well enough and my implementation is not so awful I can work on branching/merging all this stuff here on github.

Thank you,
Nikolay.

**Comment by Bajdi** [comment]
Hi Nikolay. I'm using 4 IR receivers with an Arduino Mega. I quickly came to the conclusion that it does not work. While searching for a solution I stumbled on this post. I'm very interested in the modifications you have done. Could you sent me your updated library? I would be very grateful. My email address is info [at] bajdi d0t com

**Comment by Neco777** [comment]
Bajdi, I sent you the sources. Did you receive my message?

**Comment by Bajdi** [comment]
Yes, I have just unzipped the files. I will try out the library later
today. Thank you very much :)
I'm trying to make something similar to this:
http://www.pololu.com/catalog/product/701

You can see my robots on my website: www.bajdi.com

Greetings,

Mathias

**Comment by wtip** [comment]
Could this code be posted?
I am interested in using two or more IR receivers to locate an IR beacon using a robot. Is there a way to determine the IR signal strength using this library?

**Comment by gabe0** [comment]
Can I please have a copy of this library please, Neco777?

**Comment by Neco777** [comment]
Yes, sure. Please let me know if it worked for you.
I heard from one guy that it works only for 2 receivers. For more receivers it does not work stably.

Regards,
Nikolay.

Среда, 27 марта 2013, 16:51 -07:00 от gdh37 notifications@github.com:

**Comment by Neco777** [comment]
I temporary posted the sources into my repository.
https://github.com/Neco777/IRemoteModified

**Comment by gabe0** [comment]
Awesome. I will test it and send you my results. I am working on making a smaller library since i only need the receiving part. Thank you,
Gabe.

**Comment by gabe0** [comment]
It works great with two receivers. For 3 receivers, Neco, I made _maxcount=2 but I'm sure you already knew that. It doesn't work at all with 4 though. I'm impressed with the Cpp data structure, really good work! And thanks for the help!

**Comment by hellowtisch** [comment]
When I try to use the files from Neco777's repo I receive lots of errors when trying to compile the examples from the original repo. If they don't work anymore, could anyone please add an example of using two receivers with decoding results? I guess I'm too stupid to get it to work...

**Comment by jnodev** [comment]
It's just a guess because you didn't give any examples of the errors you got, but I think you can fix the compile errors with this:
In the file IRremoteInt.h you have to replace #include < arduino.h > with #include < Arduino.h >.  

Thanks Neco777 for this lib, works like a charm.

**Comment by Pingsut** [comment]
Hi Neco777, Your lib very useful, I have a question, First time, I try 4 IR receiver on arduino UNO, It can receive data from transmitter, very good, And then I try to add more to be 5 IR receiver, It doesn't work , Could you please suggest me where I should to modify on lib file. 
Thanks Neco777

**Comment by Anadhino** [comment]
Hi Neco777 and Pingsut, Please help me.
i'm tring your lib but i have the same problem, i have 2 recivers on my arduino uno,  just one reciever work, not the second. I dont know if something is wrong, or if i have to modified something else. I copied your lib into Arduino/Librairies. Please i need some example using your lib. PLEASE HELP ME 
 this is my code :

#include <IRremote.h>
#include "U8glib.h"

int RECV_PIN = 2;
int RECV_PIN2 = 5;
int affiche=0;

IRrecv irrecv(RECV_PIN);
IRrecv irrecv2(RECV_PIN2);

decode_results results;
decode_results results2;
U8GLIB_SSD1306_128X64 u8g(U8G_I2C_OPT_NO_ACK);

void draw(int affiche) {
  // graphic commands to redraw the complete screen should be placed here  
  u8g.setFont(u8g_font_unifont);
  u8g.setPrintPos(0, 20); 
  // call procedure from base class, http://arduino.cc/en/Serial/Print
  if(affiche==0){
  u8g.print(" Jeu CETA");
  }else if(affiche==1){
          u8g.print("Recu à droite");
          }else if (affiche==2){
                    u8g.print("Recu à gauche");
                   }

}

void setup()
{
  Serial.begin(9600);
  irrecv.enableIRIn(); // Start the receiver
  irrecv2.enableIRIn(); // Start the receiver
}

void loop() {

  u8g.firstPage();  
  do {
    draw(affiche);
  } while( u8g.nextPage() );

  if (irrecv.decode(&results)) {
   // Serial.println(results.value, HEX);
   affiche=1;
    irrecv.resume(); // Receive the next value
  }

  if (irrecv2.decode(&results2)) {
   // Serial.println(results.value, HEX);
   affiche=2;
    irrecv.resume(); // Receive the next value
  }
   delayMicroseconds(500000);
}

**Comment by jnodev** [comment]
You forgot a "2" in the code for the second receiver. irrecv.resume() should be irrecv**2**.resume(), I guess. Also consider to use a [gist](https://gist.github.com/) next time you want to share a lot of code.

**Comment by Anadhino** [comment]
Thanks you soo much for your help guys it work very Well. Excuse me for the code. next time i willl add a link. excuse my english i'm a from france. THANKS.

**Comment by z3t0** [comment]
Please fork this repository and create a pull request with your changes so it can be merged here.

**Comment by Neco777** [comment]
z3t0, I have created a pull request, please have a look. Btw, I have tested it on 3 IR receivers on my Arduino Uno (I think it is Rev3) - it works. When I tried to add more instances of the receivers it stopped working. Probably because the time of reading (ISR(TIMER_INTR_NAME)) is more than 50 microseconds. I am curious if it can be solved somehow...

**Comment by khair-eddine** [comment]
hii guys, 
i'm new with IR communication and i want to know if i have to add only the modified version of the IRremote lib or add old version and the modified version by Neko777
also i would be so grateful if you gave me an example of receiving data on two receivers via two pins 
it's really urgent, i need this in my final project license. 
thanks in advance

**Comment by z3t0** [comment]
@khairov #106 is @Neco777 's  PR for adding multi receiver support but the project is currently undergoing some maintenance and it may be a while before that PR gets merged into the master branch. Since you don't have a lot of time on your hands you can contact me by email and/or skype and I can help you get setup.

As for an example on how to get multi receiver support @Neco777 could provide a sample sketch?

Skype: polarised16
Email: zetoslab@gmail.com

To get the changes made by @Neco777 I think that the best way would be to use his fork of the project for the moment. Which means downloading the zip straight from [here](https://github.com/Neco777/Arduino-IRremote) 

Feel free to contact me if you need any help

**Comment by khair-eddine** [comment]
thank you bro i will add you on skype and i really appreciate that

**Comment by z3t0** [comment]
@khairov No problem, happy to help :smile:

**Comment by z3t0** [comment]
@Neco777 My bad, I just saw the example sketch

**Comment by khair-eddine** [comment]
@z3t0  thnx for the link i download the modified version of IRremote and tryed to compile the multirecv example and it showed no error it compiles good but i didn't test it for real when i do i will tell you guys

**Comment by khair-eddine** [comment]
@Neco777 hi bro is there any solution for wiring multiple ir leds to the same arduino board ?

**Comment by Neco777** [comment]
Hello, I did not work with sending IR signals - only receiving.

**Comment by khair-eddine** [comment]
i tried multiple receiving on arduino uno and it's stable for three sensors thanks :)

**Comment by z3t0** [comment]
@khair-eddine Thanks for the feedback. How's your project going?

**Comment by khair-eddine** [comment]
@z3t0 welcome bro, it's going well

**Comment by khair-eddine** [comment]
hi guys,
 it receives the codes but it always shows the identifier of the receiver two (recv_2) :/ any idea ?

**Comment by z3t0** [comment]
@khair-eddine Could you create a gist and paste your sketch there so we can take a look. Thanks!

**Comment by khair-eddine** [comment]
it's only the IRrecvMulti example.

**Comment by z3t0** [comment]
@khair-eddine Oh okay, I'll take a look, but I do not know too much about this. @Neco777 Could you help us out?

**Comment by Neco777** [comment]
I need some time to build the example again (i mean physically).
But meanwhile, khair-eddine, can you check the code again because you say "receiver two (recv_2)", but the identifier should be "Rcv_2" if you use the example IRrecvMulti.
Serial.print("Rcv_");
Serial.print(i);
Serial.print(":");
Serial.println(btn); also, please check the example connecting 2 receivers instead of 3.

Воскресенье, 22 марта 2015, 16:25 -07:00 от Rafi Khan notifications@github.com:

**Comment by z3t0** [comment]
@Neco777 Sounds like a good plan, I will see if I can help too.

**Comment by khair-eddine** [comment]
@Neco777  sorry it's rcv not recv, i will see if i can solve the problem too.

**Comment by khair-eddine** [comment]
hii guys it works for me with 3 receivers now, when i hit a button on the remote the code figure out with three receivers

**Comment by khair-eddine** [comment]
hii @Neco777  and @z3t0  i'm using nekolay example for mutiplereceiving and i can receive on three pins but when i try to send some data on my ir led the receivers stop working and do not receive any code
it means : 
one IRsend +(  all_rcv[0] = new IRrecv(2);  rcv_count = 1;) Work.
one IRsend +(  all_rcv[0] = new IRrecv(2);all_rcv[1] = new IRrecv(2);  rcv_count = 3; or more ) do not work. 
i hope this can improve and help, i'm trying to solve this problem too

**Comment by z3t0** [comment]
@khair-eddine Sorry I have been a bit disconnected lately... but Im back! and will get to work on this! I will have a release out as soon as I manage to fix our #define error

**Comment by khair-eddine** [comment]
hii bro i can now send on multiple ir led and receive on multiple receivers, i think it's odd and it needs more work but at least it solved my problem for the moment 
what I've done is changing the TIMER_DISABLE_INTR on timer0 i don't know if that won't make other problems so when i finish the project i'll try to figure it out.

**Comment by z3t0** [comment]
@khair-eddine That sounds great. Keep us up to date on your progress and if you learn something new.

**Comment by khair-eddine** [comment]
okai bro i will :)

**Comment by nshobhit** [comment]
Hi @Neco777  ! I was able to get the library working for upto 3 IR receivers, with the use case that only 2 would be seeing signals at any time. Subsequently, I have tried integrating your approach with another working IRRemote library for Arduino Due. The approach is same, populate a global list of IRParams for each instance of receiver and and on each interrupt process them sequentially.
The code compiles fine, but does not seem to work. The state machine never reaches STOP.

Could you please take a look ?

**Comment by nshobhit** [comment]
This the Due Library
https://github.com/enternoescape/Arduino-IRremote-Due

**Comment by ipekbocegi** [comment]
@khair-eddine, @Neco777 will you please guide me how to merge the original released IRremote library and your multi transmit version and your multi receive version? I am newbie and confused within libraries. Thanks for all of your hard work. Best wishes

**Comment by khair-eddine** [comment]
@ipekbocegi  well i can send you the lib by mail if you want to use it without understanding how, well it works for me for arduino uno and only ide 1.0.5

**Comment by ipekbocegi** [comment]
@khair-eddine I would be grateful thank you

**Comment by khair-eddine** [comment]
@ipekbocegi  give me your Email adresse.

**Comment by domonation** [comment]
@khair-eddine, @Neco777 When I used the library I get the error, IRremote.h no such file or directory compilation terminated

**Comment by z3t0** [comment]
Which library? The modified one?

**Comment by domonation** [comment]
Never mind. Got it working. Using two sensors so far and it works like a dream! Getting some more in later on tonight. I have some high power LED that can handle up to 1A for really short bursts. The only problem I was having was that the IR LED stayed bright the entire time. Since I am using multiple emitters for a project, this was going to be a power problem. Using a BJT transistor as a driver, after every transmission I disabled the transistor by sending a high to pin 3.

**Comment by domonation** [comment]
Has anyone else had an issue with the timers? I am currently implementing 4 receivers but when I attempt to use the millis() function it gives me a constant value.

**Comment by domonation** [comment]
Still having issues with this timer business. Does anyone have an idea? 
Here is my code. The problem is that when I initialize my receivers I cannot use the ultrasonic sensor. It just idles after printing out the first line. Any help would be appreciated.

https://gist.github.com/aa4d5d9135c809662954.git

**Comment by AnalysIR** [comment]
Did you consider, just using OR gates to feed all of the 4 receivers into one pin?

**Comment by z3t0** [comment]
@domonation the gist does not exist

**Comment by z3t0** [comment]
@domonation Could you please create a new issue to keep things easily maintainable. Thank you.

**Comment by domonation** [comment]
@AnalysIR I considered using OR gates but I need to be able to distinguish between unique emitter IDs. 

https://gist.githubusercontent.com/domonation/aa4d5d9135c809662954/raw/88b841a8f2979c616f39408e1eb09fbad3fb74c9/irbeaconreceiver

**Comment by domonation** [comment]
I started a new issue @z3t0 
https://github.com/shirriff/Arduino-IRremote/issues/180

**Comment by z3t0** [comment]
@domonation Sorry just saw the sketch. Will read and report back on the other issue

**Comment by mojobi1** [comment]
@khair-eddine, following @khair-eddine, will you please guide me how to merge the original released IRremote library and your multi transmit version and your multi receive version? I am newbie and confused within libraries. Thanks for all of your hard work. Best wishes

**Comment by kwdaily** [comment]
So I'm trying to test my multiple receivers for a IR beacon project, and unless I unplug one of the sensors, the Serial Monitor shows me that both sensors are triggered everytime I press a button on the remote, even with directional cones mounted on both sensors, and baffles separating them. My code is as follows (taken from the IRrecvMulti example.

#include "IRremote.h"
#include "CppList.h"

bool _initialized = false;

int rcv_count;
IRrecv **all_rcv;

void setup() {
  if (_initialized) return;

  Serial.begin(9600);

  rcv_count = 2;
  all_rcv = (IRrecv *_)malloc(rcv_count_sizeof(int));
  all_rcv[0] = new IRrecv(2);
  all_rcv[1] = new IRrecv(3);
  //all_rcv[2] = new IRrecv(4);
 // all_rcv[3] = new IRrecv(5);

  for (int i=0; i<rcv_count; ++i){
    all_rcv[i]->enableIRIn();
  }

  _initialized = true;
}

void loop() {
  for (int i=0; i<rcv_count; ++i){
    decode_results results;
    if (all_rcv[i]->decode(&results)) {
      Serial.print("Rcv_");
      Serial.print(i);
      Serial.println("");
      all_rcv[i]->resume();
    }
  }
}

**Comment by AnalysIR** [comment]
In common with visible light IR will reflect off surfaces/walls etc and reach both receivers unless the physical design is 100%.

**Comment by etix** [comment]
For my personal needs I've implemented the [multi-receiver support](https://github.com/etix/Arduino-IRremote/commit/23caf5a832d05cfaf87fe8b792c5caf83b51afe8) into [my fork](https://github.com/etix/Arduino-IRremote) based on the latest IRremote master. I'm using it successfully with 5 receivers on a Teensy LC (48mhz, so fast enough for the job). It's a quick and dirty hack done in half an hour, use it at your own risk :-)

Note: I've hardcoded the maximum number of receivers to 5, see the commit for details.

**Comment by kwdaily** [comment]
Awesome, Thanks! Sorry for the late response, this email was hiding! I'll
definitely check this out!

On Sat, Jan 2, 2016 at 3:01 PM, Ludovic Fauvet notifications@github.com
wrote:

> For my personal needs I've implemented the multi-receiver support
> https://github.com/etix/Arduino-IRremote/commit/23caf5a832d05cfaf87fe8b792c5caf83b51afe8
> into my fork https://github.com/etix/Arduino-IRremote based on the
> latest IRremote master. I'm using it successfully with 5 receivers on a
> Teensy LC (48mhz, so fast enough for the job). It's a quick and dirty hack
> done in half an hour, use it at your own risk :-)
> 
> Note: I've hardcoded the maximum number of receivers to 5, see the commit
> for details.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/z3t0/Arduino-IRremote/issues/20#issuecomment-168427956
> .

## 

Kenneth Daily
kwdaily@gmail.com

**Comment by gauravgcchawla** [comment]
#include "CppList.h"

```
                 ^
```

compilation terminated.

exit status 1
Error compiling.

i am getting error can someone help me with it

**Comment by etix** [comment]
Give a try to my version first since it doesn't require CppList and it's based on a recent master. How many receivers do you need? On what hardware?

**Comment by gauravgcchawla** [comment]
2 receivers to controle multiple relay and i am using arduino uno

**Comment by AnalysIR** [comment]
Why do you need 2 receivers? Can you just not use 2 different signal codes - one for each relay?

**Comment by gauravgcchawla** [comment]
I want one of the receiver on 1st floor of my house

**Comment by gauravgcchawla** [comment]
@etix
i am using your IRlibrary but still my program is not working can u please help

#include 'IRremote.h'
#include 'IRremoteInt.h'

int relay_pin1 = 13;
int recv_pin = 11;
int recv_pin2 = 12;
int relay_pin2 = 4;
int relay_pin3 = 5;
int relay_pin4 = 6;
int actual_state = HIGH;

IRrecv irrecv(recv_pin);
IRrecv irrecvs(recv_pin2);
decode_results results;

void setup() {
    pinMode(relay_pin1, OUTPUT);
    pinMode(relay_pin2, OUTPUT);

```
pinMode(relay_pin3, OUTPUT);
pinMode(relay_pin4, OUTPUT);
Serial.begin(9600);
irrecv.enableIRIn(); // Start the receiver
```

irrecvs.enableIRLn();

}

void loop() {
    if (irrecv.decode(&results)) {
if(irrecvs.decode(&results)){
        Serial.println(results.value, HEX);
        if (results.value == 0x80BFC13E) { // On/Off button
            actual_state = (actual_state == LOW) ? HIGH : LOW;
            digitalWrite(relay_pin1, actual_state);
        }
        if (results.value == 0x80BF5BA4) { // On/Off button
            actual_state = (actual_state == LOW) ? HIGH : LOW;
            digitalWrite(relay_pin2, actual_state);
        }
         if (results.value == 0x80BFB34C) { // On/Off button
            actual_state = (actual_state == LOW) ? HIGH : LOW;
            digitalWrite(relay_pin3, actual_state);
         }
            if (results.value == 0x80BF5BA4) { // On/Off button
            actual_state = (actual_state == LOW) ? HIGH : LOW;
            digitalWrite(relay_pin2, actual_state);
        }
         if (results.value == 0x80BFB34C) { // On/Off button
            actual_state = (actual_state == LOW) ? HIGH : LOW;
            digitalWrite(relay_pin3, actual_state);
        }
         if (results.value == 0x80BF03FC) { // On/Off button
            actual_state = (actual_state == LOW) ? HIGH : LOW;
            digitalWrite(relay_pin4, actual_state);
        }
        irrecv.resume(); // Receive the next value
irrecvs.resume();
    }
}

**Comment by ElectricRCAircraftGuy** [comment]
If only one receiver is expected to be used at a time, this could be solved with hardware. Simply use an "AND" gate to "AND" the signals together from multiple IR receivers, and use a single IRrecv object, and Arduino pin, to decode. (You would _not_ use an OR gate here since the signals from the IR receivers are active LOW.) AND gates that would work include: CD4081, CD4082, 74HC08, and SN74LS21, etc. The 74HC08s, for example, are very cheap on Ebay.

#### Recommendation to IRremote library:

-add this info in the readme, so people know they can have receivers in multiple locations all "AND"ed together.

**Comment by bengtmartensson** [comment]
> Simply use an "AND" gate to "AND" the signals together from multiple IR receivers,

Actually, I think it is even easier: Looking at the data sheet for TSOP344\* reveals that it uses an NPN transistor as output device, with just a 30k resistor to plus. I.e., it is almost "open collector", and it should be possible to just connect the output pins of several devices together. This makes up the "AND gate". The max current is 5 mA, so theoretically it should be possible, with 5V supply voltage  to connect  0.005/(5/30000) = 30 receivers together. 

I have not tried it myself, though :-).

**Comment by z3t0** [comment]
@ElectricRCAircraftGuy 

> -add this info in the readme, so people know they can have receivers in multiple locations all "AND"ed together.

Will do thanks!

Also would you be interested in becoming a collaborator on this project? Your work is wonderful and you also seem to be very knowledgeable and active.

**Comment by gauravgcchawla** [comment]
it worked can someone tell me is it possible to use bluetooth and ir at the same time

**Comment by pcoughlin** [comment]
Bluetooth is Radio Frequency (RF), IR is Infra-Red light.  They do not conflict in any manner in their spectrum use. The hardware required for RF and the hardware for IR is a completely different set of additional parts or additional boards that are used in connection with the Processor board.

Please Google "[arduino bluetooth radio](https://www.google.com/search?q=arduino+bluetooth+radio&oq=arduino+bluetooth+radio&aqs=chrome..69i57&sourceid=chrome&es_sm=122&ie=UTF-8)" to see more about bluetooth.

**Comment by ricycc** [comment]
Hey guys,I'm using 4 IR receivers on Arduino mega 2560 ADK board.I merged the libraries from @z3t0 and @Neco777 ,but when compling the IRrecvMulti sample,the IDE has warning like the following:
Arduino:1.6.8 (Windows 10),"Arduino Mega ADK"

~\Arduino\libraries\IRremote\irRecv.cpp: In member function 'int IRrecv::decode(decode_results*)':

~\Arduino\libraries\IRremote\irRecv.cpp:14:11: error: 'class decode_results' has no member named 'overflow'

  results->overflow = irparams.overflow;

```
       ^
```

~\Arduino\libraries\IRremote\irRecv.cpp:14:31: error: 'struct irparams_t' has no member named 'overflow'

  results->overflow = irparams.overflow;

```
                           ^
```

~\Arduino\libraries\IRremote\irRecv.cpp: At global scope:

~\Arduino\libraries\IRremote\irRecv.cpp:105:1: error: prototype for 'IRrecv::IRrecv(int, int)' does not match any in class 'IRrecv'

 IRrecv::IRrecv (int recvpin, int blinkpin)

 ^

In file included from ~\Arduino\libraries\IRremote\irRecv.cpp:1:0:

~\Arduino\libraries\IRremote\IRremote.h:77:7: error: candidates are: constexpr IRrecv::IRrecv(const IRrecv&)

 class IRrecv

```
   ^
```

~\Arduino\libraries\IRremote\irRecv.cpp:99:1: error:                 IRrecv::IRrecv(int)

 IRrecv::IRrecv (int recvpin)

 ^

~\Arduino\libraries\IRremote\irRecv.cpp:154:24: error: no 'bool IRrecv::isIdle()' member function declared in class 'IRrecv'

 bool  IRrecv::isIdle ( )

```
                    ^
```

exit status 1
Error compiling for board Arduino Mega ADK.

I've changed to complie for Arduino UNO,it just stay the same.
I've also tried the librariy from @etix ,it worked well with four IR receivers,but just wont work with an UltraSonic sensor,I think there might be some problem with Timer,just not for sure.
I'm still starter on this.
Could you possibly tell me what caused this?

**Comment by p2mini** [comment]
Hi everyone,i am new to arduino coding field. For my final year project work i am using 6 ir receivers. But, facing problem in it.... Plz Plz can anyone help me as soon as possible.
Waiting for reply
Thnxs.

**Comment by ricycc** [comment]
@p2mini ,I was dealing with 4 IR receivers two weeks ago,and sucessed by using @etix 's library.They worked together.Just need a period longer than 25 milliseconds between each other.

**Comment by p2mini** [comment]
@ricycc Thnxs a lot for ur reply.. Can i get the sketch wht u have used for 4 ir receivers? If u don't mind. I have one more doubt only by using this library in my sketch will it work... ?? Plz help me..

**Comment by p2mini** [comment]
Hi, Thnxs for guiding me.. Its working only for 4 ir receivers... Happy for that.. But my work for 6 ir receivers.. Plx guide me..

**Comment by p2mini** [comment]
@khair-eddine can i get library for multiple ir receivers.. As i have used etix's library but its working only for 4 ir receivers. 
Waiting for your reply.
Thxs in advance.

**Comment by khair-eddine** [comment]
hello bros sorry for being absent during all this time, this is an old post
and the library wich i modified is not the same as now it was updated with
a lot of enhancements but without solving the problem of multiple
transmitters on same board this is the library that i modified a year ago i
think it should be work for you or all you must do is to replace it by the
new one from github.

2016-06-15 12:35 GMT+02:00 p2mini notifications@github.com:

> @khair-eddine https://github.com/khair-eddine can i get library for
> multiple ir receivers.. As i have used etix's library but its working only
> for 4 ir receivers.
> Waiting for your reply.
> Thxs in advance.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/z3t0/Arduino-IRremote/issues/20#issuecomment-226150291,
> or mute the thread
> https://github.com/notifications/unsubscribe/AK9cKZQF-zRDofrJodaC1auBTfmU-ZoYks5qL9VzgaJpZM4APAGG
> .

## 

Cordialement,
Khaireddine Hlali

**Comment by p2mini** [comment]
@khair-eddine Hi bro thnxs for ur reply.. it means a lot for me.. Sorry to say m nt getting d new modified library about wch u r taking.. Plzz can u email me the multiple ir receiver library... 
Waiting for ur reply..
Thnxs in advance.

**Comment by khair-eddine** [comment]
@p2mini  yes bro write ur email adress ill email u

**Comment by p2mini** [comment]
@khair-eddine thnxs for ur reply.. here s my email id.. pranshu.789@yahoo.com
Waitng for ur mail.. Thnxs again..

**Comment by lemorlenny** [comment]
I used an 4051 hardware multiplexer to connect 8 TSOP using 3+1 Arduino pins but I have problems about emitter/receiver synchronization.
https://www.google.it/search?q=4051+multiplexer&safe=off&tbm=isch&tbo=u&source=univ&sa=X&ved=0ahUKEwjW6KWR5LbNAhWEVRoKHZrpAn8QsAQIMA&biw=1920&bih=995#imgrc=HOzSOeBX4r7NPM%3A

**Comment by paul760727** [comment]
@etix  Hi  I used your IRremote library  I successful work for 5 ir receivers in arduino uno but it work very lag so I buy arduino due. When I used due the library can't use.  Plz help me

**Comment by etix** [comment]
@paul760727 I don't own an Arduino Due, as told earlier in the thread I only tested the code on a Teensy LC.

**Comment by akame99** [comment]
did someone here sucessfully connected 6 IR receivers on their MCU? can someone please help me.

**Comment by Emperor-Sultan** [comment]
Hey all. z3t0 I've been on one of your other threads pertaining to ir receivers, and I had my problem solved. Thanks again.
I know this is an old topic from over 4 years ago, but maybe someone can help.
I downloaded Neco's modified ir, and it compiles like everyone says, but when I try to cut and paste some from his into my code, it doesn't work. I've always had one ir working, but the 2nd one is a complete nightmare that I come back to after work every night. I'm trying to just work with 2 ir receivers that are signaled by 2 separate emitters from a separate source and are not connected to the arduino nano that I'm using).

I would like to email someone my different versions of sketches I've put together, if anykne is willing to tackle this and point me in the right direction. I just dont want to post it for everyone to see, as I am working on something private and can't really share with anyone. Sorry

**Comment by z3t0** [comment]
If it's a private project than feel free to shoot me an email

**Comment by Emperor-Sultan** [comment]
Rafi Khan,


Thanks for the reply. Attached are [x3] files with different libraries included to try and make it work. One file has many libraries, and still nothing. All 3 sketches compile, its just not working as I need it to. Only first ir detector is picking up signal. The 2nd ir detector is not responding to the signals. The signals are coming from 2 separate ir transmitters that are grounded and powered from separate batteries.


I have 10k resistors on both detectors, and 330 ohm resistor on the transmitters.


These are plain 2 pin ir transmitters and detectors. I may need to use the TSOP detectors instead. But those are 3 pins, and I'm trying to keep it as simple as possible.


I have swapped the wiring for the 2 detectors to see if the detector is dead, but they both work fine. I strongly believe it is either the code, or for some reason others can use more than 1 ir detector, but I cant. BTW, I have used an Arduino pro trinket 5v, and an Arduino nano.



Please don't share my work, as its confidential. And reply back to me via email.


Thank you,

Abdul Hotaki



________________________________

**Comment by z3t0** [comment]
Hi. Can you please try emailing me again. I did not receive anything yet.

**Comment by AldenNg** [comment]
Hi All,

I am new to this thread.

I am currently working on a project which requires me to build a circuit using wireless technology. Size is crucial factors for this project.

The scope in general will be as such:

1. One transmitter needs to connect to 5 receivers (Complete set)
2. Upon receiving signal from transmitter, each receiver is to light up a light bulb in a "blinking" mode
3. The signal of this complete set should not interfere into other sets. 

Anyone can advise the most economical manner to build this as i am on a tight budget.

Thanks in advance.
Alden

**Comment by AnalysIR** [comment]
> Anyone can advise the most economical manner to build this as i am on a tight budget.

Just buy 6 Arduino Nano clones from Aliexpress for circa $2 each incl shipping and use one for each receiver/transmitter. As a bonus you get an LED with each unit to blink.

**Regarding the more technical aspects**, you have not provided enough detailed information and perhaps you should consider creating a new thread and posting details about the effort/steps you have taken so far.

**Comment by z3t0** [comment]
This conversation has been locked for non-contributors to encourage new threads to be created for separate issues. Also, to point out that this feature is experimental and not a part of the main code base.

Although, I have thought about merging this upstream I decided against it. The feature is advanced enough that those needing it can be expected to use parts of the library to easily create what they need. However, maintaining and creating a more generic implementation for all the supported boards is far less trivial and also not entirely within the scope of this library.

**Comment by ArminJo** [comment]
I documented it in the readme, so this PR ist no longer needed.
`You can use **multiple IR receiver** by just connecting the output pins of several devices together. The IR receivers use an NPN transistor as output device with just a 30k resistor to VCC. This is almost "open collector" and allows connecting of sereral output pins.`
